src/infill/subDivCube.h

```diff
@@ -0,0 +1,81 @@
+#ifndef INFILL_SUBDIVCUBE_H
+#define INFILL_SUBDIVCUBE_H
+
+#include "../sliceDataStorage.h"
+
+namespace cura
+{
+
+class Infill;
+
+class SubDivCube
+{
+public:
+    /*!
+     * Constructor for SubDivCube. Recursively calls itself eight times to flesh out the octree.
+     * \param mesh contains infill layer data and settings
+     * \param my_center the center of the cube
+     * \param depth the recursion depth of the cube (0 is most recursed)
+     */
+    SubDivCube(SliceMeshStorage& mesh, Point3& center, int depth);
+
+    ~SubDivCube(); //!< destructor (also destroys children
+
+    /*!
+     * Precompute the octree of subdivided cubes
+     * \param mesh contains infill layer data and settings
+     */
+    static void precomputeOctree(SliceMeshStorage& mesh);
+    /*!
+     * Generates the lines of subdivision of the specific cube at the specific layer. It recursively calls itself, so it ends up drawing all the subdivision lines of sub-cubes too.
+     * \param z the specified layer height
+     * \param result (output) The resulting lines
+     * \param directional_line_groups Should be nullptr. Used internally to keep track of line segments that are all pointing the same direction for line segment combining
```

Comment:
The typical way to guard against this is to make a public function that has no `directional_line_groups` parameter, which calls a private recursive function that has the parameter. You won't need all that hassle with the `calloc` and freeing the memory then.